topics/tutorials/postfix_completion/advanced_postfix_templates.md

```diff
@@ -0,0 +1,93 @@
+[`PostfixTemplateWithExpressionSelector`](upsource:///platform/lang-impl/src/com/intellij/codeInsight/template/postfix/templates/PostfixTemplateWithExpressionSelector.java)
+and
+[`PostfixTemplateExpressionSelector`](upsource:///platform/lang-impl/src/com/intellij/codeInsight/template/postfix/templates/PostfixTemplateExpressionSelector.java)
+classes.
+
+**Example:**
+[`IntroduceFieldPostfixTemplate`](upsource:///java/java-impl/src/com/intellij/codeInsight/template/postfix/templates/IntroduceFieldPostfixTemplate.java)
+introduces a Java class field, allowing to select one of the non-void expressions at the current offset.
+
+## Live Template Syntax-Based Postfix Templates
+
+The IntelliJ Platform-based IDEs provide the [Live Templates](https://www.jetbrains.com/help/idea/using-live-templates.html) feature.
+It allows defining a template text with dynamic expressions, replaced with actual values depending on the context.
+If the implemented postfix template's expanding behavior can be achieved with the live template syntax, it is much easier to extend the
+[`StringBasedPostfixTemplate`](upsource:///platform/lang-impl/src/com/intellij/codeInsight/template/postfix/templates/StringBasedPostfixTemplate.java)
+than implementing the expansion behavior programmatically.
+
+**Example:** [`StreamPostfixTemplate`](upsource:///java/java-impl/src/com/intellij/codeInsight/template/postfix/templates/StreamPostfixTemplate.java) wraps array expression within the `Arrays.stream()` method.
+
+> See the [](live_templates.md) section in SDK Docs for information on implementing the live templates feature in a plugin.
+>
+{type="note"}
+
+## Editable Postfix Templates
+
+All postfix templates that return true from the
+[`PostfixTemplate.isEditable()`](upsource:///platform/lang-impl/src/com/intellij/codeInsight/template/postfix/templates/PostfixTemplate.java)
+method and have a key starting with `.` (dot) can be edited.
+In the simplest case, only the template's name can be edited.
+To provide more advanced editing possibilities, like creating new templates in the settings dialog, editing the template's content, variables, expression conditions, etc., a plugin must implement the related
+[`PostfixTemplateProvider's`](upsource:///platform/lang-impl/src/com/intellij/codeInsight/template/postfix/templates/PostfixTemplateProvider.java)
+methods:
+
+- `PostfixTemplateProvider.createEditor()` - returns an instance of the
+  [`PostfixTemplateEditor`](upsource:///platform/lang-impl/src/com/intellij/codeInsight/template/postfix/templates/editable/PostfixTemplateEditor.java)
+  class which is responsible for building the UI settings for a particular template and creating a template from the settings provided in this UI.
+  The editor UI may contain settings controls other than the textual editor, like lists, checkboxes, etc.
+- `PostfixTemplateProvider.writeExternalTemplate()`/`readExternalTemplate()` - serializes/deserializes a given template to/from XML elements.
```

Comment:
unfortunate line breaks for me,split explicitly

Reply:
I follow the convention to put a single sentence per line + @halirutan's idea to put links in separate lines as well, which I like. I think we should agree on a consistent approach.

Reply:
Remove the `/` and write an "and" there. Let's see if it fixes it.

Reply:
Sorry, I overlooked it. Fixed, thanks.